XenModel/Actions/ParallelAction.cs

```diff
@@ -44,58 +44,128 @@ namespace XenAdmin.Actions
+        /// </summary>
+        public ParallelAction(string title, string startDescription, string endDescription, List<AsyncAction> subActions, bool suppressHistory, bool showSubActionsDetails, int maxNumberOfParallelActions = DEFAULT_MAX_NUMBER_OF_PARALLEL_ACTIONS)
+            : base(null, title, startDescription, endDescription, subActions, suppressHistory, showSubActionsDetails)
         {
-            _queuePC = new ProduceConsumerQueue(Math.Min(maxNumberOfParallelActions, subActions.Count));
+            GroupActionsByConnection();
+            CreateQueues(maxNumberOfParallelActions);
         }
 
-        protected override void RunSubActions(List<Exception> exceptions)
+        public ParallelAction(string title, string startDescription, string endDescription, List<AsyncAction> subActions, int maxNumberOfParallelActions = DEFAULT_MAX_NUMBER_OF_PARALLEL_ACTIONS)
+            : this(title, startDescription, endDescription, subActions, false, false, maxNumberOfParallelActions)
+        { }
+
+        private void GroupActionsByConnection()
         {
-            foreach (AsyncAction subAction in subActions)
+            foreach (AsyncAction action in subActions)
             {
-                AsyncAction action = subAction;
-                action.Completed += action_Completed;
-                _queuePC.EnqueueItem(
-                    () =>
+                if (action.Connection != null)
+                {
+                    if (action.Connection.IsConnected)
                     {
-                        try
+                        if (!actionsByConnection.ContainsKey(action.Connection))
                         {
-                            action.RunExternal(Session);
+                            actionsByConnection.Add(action.Connection, new List<AsyncAction>());
                         }
-                        catch (Exception e)
-                        {
-                            Failure f = e as Failure;
-                            if (f != null && Connection != null &&
-                                f.ErrorDescription[0] == Failure.RBAC_PERMISSION_DENIED)
-                            {
-                                Failure.ParseRBACFailure(f, Connection, Session ?? Connection.Session);
-                            }
-                            exceptions.Add(e);
-                            // Record the first exception we come to. Though later if there are more than one we will replace this with non specific one.
-                            if (Exception == null)
-                                Exception = e;
-                        }
-                    });
 
+                        actionsByConnection[action.Connection].Add(action);
+                    }
+                }
+                else
+                {
+                    actionsWithNoConnection.Add(action);
+                }
+            }
+        }
+
+        private void CreateQueues(int maxNumberOfParallelActions)

```

Comment:
I don't think this CreateQueues() function is needed, as you do all the work again in RunSubActions()